proposals/mmmm-async-let.md

```diff
@@ -0,0 +1,733 @@
+```
+
+This is a simple rule and allows us to bring the feature forward already. It might be possible to employ control flow based analysis to enable "only the first reference to the specific `async let` on each control flow path has to be an `await`", as technically speaking, every following await will be a no-op and will not suspend as the value is already completed, and the placeholder has been filled in.
+
+### Implicit `async let` awaiting 
+
+A `async let` that was declared but never awaited on *explicitly* as the scope in which it was declared exits, will be awaited on implicitly. These semantics are put in place to uphold the Structured Concurrency guarantees provided by `async let`.
+
+To showcase these semantics, let us have a look at this function which spawns two child tasks, `fast` and `slow` but does not await on any of them:
+
+```swift
+func go() async { 
+  async let f = fast() // 300ms
+  async let s = slow() // 3seconds
+  return "nevermind..."
+  // implicitly: cancels f
+  // implicitly: cancels s
+  // implicitly: await f
+  // implicitly: await s
+}
+```
+
+Assuming the execution times of `fast()` and `slow()` are as the comments next to them explain, the `go()` function will _always_ take at least 3 seconds to execute. Or to state the rule more generally, any structured invocation will take as much time to return as the longest of its child tasks takes to complete.
+
+As we return from the `go()` function without ever having awaited on the `f` or `s` values, both of them will be implicitly cancelled and awaited on before returning from the function `go()`. This is the very nature of structured concurrency, and avoiding this can _only_ be done by creating non-child tasks, e.g. by using `detach` or other future APIs which would allow creation of non-child tasks.
+
+If we instead awaited on one of the values, e.g. the fast one (`f`) the emitted code would not need to implicitly cancel or await it, as this was already taken care of explicitly:
+
+```swift
+func go2() async {
+  async let f = fast()
+  async let s = slow()
+  _ = await f
+  return "nevermind..."
+  // implicitly: cancels s
+  // implicitly: awaits s
+}
+```
+
+The duration of the `go2()` call remains the same, it is always `time(go2) == max(time(f), time(s))`.
+
+Special attention needs to be given to the `async let _ = ...` form of declarations. This form is interesting because it creates a child-task of the right-hand-side initializer, however it actively chooses to ignore the result. Such a declaration (and the associated child-task) will run and be awaited-on implicitly, as the scope it was declared in is about to exit — the same way as an unused `async let` declaration would be.
+
+> It may be interesting for a future proposal to explore the viability of sugar to `spawn voidReturningFunction()` directly, as a `Void` returning function may often not necessarily want to be awaited on for control-flow reasons, as variables of interesting types would be.
```

Comment:
Does `spawn voidReturningFunction()` need to change?

Reply:
let me remove this actually, it's superseded by `async {}` which is an update to structured concurrency in https://github.com/apple/swift-evolution/pull/1345